### DISCUSSION TITLE
Model textures are too big and cause crashes

### DISCUSSION THREAD

**Discussion by DavidWeng916** [pixiv/vroid-sdk-developers#107] [Q&A]
Hello, I have a question, I used DefaultApi.GetStaffPicksAsync() API to get the recommendation list, there is one model in the recommended list => [https://hub.vroid.com/en/characters/2421223126683964798/models/6792752807926138160](https://hub.vroid.com/en/characters/2421223126683964798/models/6792752807926138160), my iPhone XR will crash when I create this model, after tracking down this model. I found out that its texture uses **six 4096x4096** textures, which takes up a lot of memory. I tried to use the attribute **original_file_size** to filter the model, but this attribute is only **19621684** which is not particularly large, so I guess it doesn't seem to be included in the texture size part of the attribute. I would like to ask if there is any suggested way to exclude models that are too big for the texture.

Attached Unity Profiler screenshot
<img width="448" alt="截圖 2023-08-16 下午11 07 45" src="https://github.com/pixiv/vroid-sdk-developers/assets/115971193/6852137b-8ce8-4d3e-892f-98211e7834a8">

**Comment by nyamadan** [comment]
I'm sorry for the late reply.

The original_file_size represents the file size, which is different from the runtime size.
For example, a PNG file is decompressed at runtime as 32-bit RGBA, which is Width x Height x 4 bytes.

We have not been able to find an easy way to calculate this.

UniVRM can implement an ITextureDeserializer to deserialize textures, so it may be possible to calculate the file size after decompression.

```cs
public class TextureDeserializer : ITextureDeserializer
    {
        public int Threshold = ...;
        public int RuntimeTextureSize = 0;
        public bool Oversize = false;

        public async Task<Texture2D> LoadTextureAsync(DeserializingTextureInfo textureInfo, IAwaitCaller awaitCaller)
        {
            switch (textureInfo.DataMimeType)
            {
                case "image/png":
                    var width =  (textureInfo.ImageData[16] << 24) |
                                 (textureInfo.ImageData[17] << 16) |
                                 (textureInfo.ImageData[18] <<  8) |
                                 textureInfo.ImageData[19];
                    var height = (textureInfo.ImageData[20] << 24) |
                                 (textureInfo.ImageData[21] << 16) |
                                 (textureInfo.ImageData[22] <<  8) |
                                 textureInfo.ImageData[23];
                    var size = width * height * 8;
                    Debug.Log($"{width}x{height}x32 = ${size} bytes");
                    this.RuntimeSize += size;
                    if(RuntimeSize > Threshold) {
                      Oversize = true;
                      // return error texture.
                      return new Texture2D(2, 2);
                    }
                    break;
                case "image/jpeg":
                    // load width and height, same as above.
                    break;
            }
        }
    }

// Load VRM like this
var textureDeserializer = new TextureDeserializer();
Vrm10.LoadBytesAsync(..., textureDeserializer: textureDeserializer)
if(textureDeserializer.Oversize) {
  Debug.Log("Failed to load VRM!"); // avoid crashes
}
    
```

**Reply by DavidWeng916** [reply]
Hi, @nyamadan I tried your suggestion but the logic seems to be in the character-building process. When `RuntimeTextureSize` reaches our defined limit, it will return 2x2 `Texture2D` directly, but that would make the character look weird like it's broken. So we don't currently have any way to know the correct size of the model before we build it, right? Because I want to know the size of the model first, and remove the larger model from the model list, so that players can avoid crashing when they select a larger model and run out of memory. Also, I've tried other apps that use VRoid Hub and your Example and it also causes crashes. Do you have a solution for this problem in the future? Or maybe consider not having size oversized characters put on the official recommendation list?
